src/geojson_types.jl

```diff
@@ -0,0 +1,350 @@
+    end
+end
+
+function Base.iterate(fc::AbstractFeatureCollection{D,T}, state=1) where {D,T}
+    (1 <= state <= length(fc)) || return nothing
+    val = fc[state]::Feature{D,T}
+    return val, state + 1
+end
+
+Base.show(io::IO, fc::FeatureCollection) = print(io, "FeatureCollection with $(length(fc)) Features")
+Base.eltype(::Type{<:AbstractFeatureCollection{D,T}}) where {D,T} = Feature{D,T}
+Base.IteratorEltype(::Type{<:AbstractFeatureCollection}) = Base.HasEltype()
+Base.length(fc::AbstractFeatureCollection) = length(features(fc))
+Base.lastindex(fc::AbstractFeatureCollection) = length(features(fc))
+Base.IteratorSize(::Type{<:AbstractFeatureCollection}) = Base.HasLength()
+Base.size(fc::AbstractFeatureCollection) = size(features(fc))
+Base.getindex(fc::AbstractFeatureCollection, i::Int) = features(fc)[i]
+Base.IndexStyle(::AbstractFeatureCollection) = IndexLinear()
+
+bbox(x::GeoJSONT) = getfield(x, :bbox)
+
+"""
+    LazyFeatureCollection{D,T}(bbox::Union{Nothing,Vector{T}}, features::Vector{LazyFeature{D,T}}, crs::Union{Nothing,String})
+
+A FeatureCollection with `D` dimensional geometry in its features, but it's features are
+lazily parsed from the GeoJSON string. Indexing into the collection will parse the feature.
+This can be more efficient when interested in only a few features from a large collection,
+or parsing a very large collection iteratively without loading it all into memory.
+"""
+struct LazyFeatureCollection{D,T} <: AbstractFeatureCollection{D,T}
+    bbox::Union{Nothing,Vector{T}}
+    features::Vector{LazyFeature{D,T}}
+    crs::Union{Nothing,CRS}
+    # TODO Use features::JSON3.Array directly once we can parse the tape to find the
+    # feature offsets in the buffer?
+end
+features(fc::LazyFeatureCollection) = collect(fc.features)
+
+Base.show(io::IO, x::LazyFeatureCollection) = print(io, "LazyFeatureCollection with $(length(x.features)) features")
+Base.getindex(x::LazyFeatureCollection{D,T}, i::Int) where {D,T} = JSON3.read(codeunits(x.features[i]), Feature{D,T})::Feature{D,T}
+
+struct GeoJSONWrapper{D,T,X<:GeoJSONT{D,T}}
+    obj::X
+end
+GeoJSONWrapper{D,T}(obj::X) where {D,T,X<:GeoJSONT{D,T}} = GeoJSONWrapper{D,T,X}(obj)
+
+# symbol (from json string type) to struct mapping
+function geom_mapping(D, T)
+    (;
+        Point=Point{D,T},
+        MultiPoint=MultiPoint{D,T},
+        LineString=LineString{D,T},
+        MultiLineString=MultiLineString{D,T},
+        Polygon=Polygon{D,T},
+        MultiPolygon=MultiPolygon{D,T},
+        GeometryCollection=GeometryCollection{D,T}
+    )
+end
+function obj_mapping(D, T)
+    (;
+        Feature=Feature{D,T},
+        FeatureCollection=FeatureCollection{D,T}
+    )
+end
+@inline StructTypes.StructType(::Type{<:GeoJSONWrapper}) = StructTypes.CustomStruct()
+@inline StructTypes.lower(x::GeoJSONWrapper) = x.obj
+@inline StructTypes.lowertype(::Type{<:GeoJSONWrapper{D,T}}) where {D,T} = GeoJSONT{D,T}
+
+const point = "Point"
+@inline typestring(::Type{<:Point}) = point
```

Comment:
I doubt these `@inline` macros are required, this will compile away. 

And is there a reason to not just return the string from the function?